test/rules-custom.spec.ts

```diff
@@ -1,72 +1,87 @@
-      eVar11: 'test',
-      eVar22: 'test',
-    };
+      afterEach(async () => {
+        await testHarness.tearDown();
+      });
 
-    setupGlobals(([
-      ['s', s],
-    ]));
+      after(async () => {
+        await testEnv.tearDown();
+      });
 
-    const rules: DataLayerRule[] = [
-      {
-        source: 's[^(eVar1)]',
-        operators: [
+      it('only triggers rules for property changes that match rule selector', async () => {
+        const rules: DataLayerRule[] = [
           {
-            name: 'insert',
-            value: 'eVar1',
+            source: 's[^(eVar1)]',
+            operators: [
+              {
+                name: 'insert',
+                value: 'eVar1',
+              },
+            ],
+            destination: 'FS.event',
           },
-        ],
-        destination: 'FS.event',
-      },
-      {
-        source: 's[^(eVar2)]',
-        operators: [
           {
-            name: 'insert',
-            value: 'eVar2',
+            source: 's[^(eVar2)]',
+            operators: [
+              {
+                name: 'insert',
+                value: 'eVar2',
+              },
+            ],
+            destination: 'FS.event',
           },
-        ],
-        destination: 'FS.event',
-      },
-    ];
+        ];
 
-    ExpectObserver.getInstance().create({ rules });
+        testHarness = await testEnv.createTestHarness(rules, {
+          s: {
+            eVar11: 'test',
+            eVar22: 'test',
+          },
+        });
 
-    expectGlobal('s').eVar11 = 11;
+        await testHarness.execute(() => {
+          globalThis.s.eVar11 = '11';
+        });
 
-    // Wait for event dispatch to result in rule destination calls
-    let [eventName, payload] = await waitForFS('event');
-    expectEqual('eVar1', eventName);
-    expectEqual(11, payload.eVar11);
+        // Wait for event dispatch to result in rule destination calls
+        let [eventName, payload] = await testHarness.popEvent();
+        expectEqual('eVar1', eventName);
+        expectEqual('11', payload.eVar11);
 
-    expectGlobal('s').eVar22 = 22;
+        await testHarness.execute(() => {
+          globalThis.s.eVar22 = '22';
+        });
 
-    [eventName, payload] = await waitForFS('event');
-    expectEqual('eVar2', eventName);
-    expectEqual(22, payload.eVar22);
+        [eventName, payload] = await testHarness.popEvent();
+        expectEqual('eVar2', eventName);
+        expectEqual('22', payload.eVar22);
 
-    // Verify no other calls were made to rule destinations
-    await new Promise<void>((resolve) => {
-      setTimeout(() => {
-        expectNoCalls(expectGlobal('FS'), 'event');
-        resolve();
-      }, 1000);
+        // Verify no other calls were made to rule destinations
+        await new Promise<void>((resolve, reject) => {
+          testHarness.popEvent(500)
+            .then(() => {
+              reject(new Error('Expected rejected promise due to no FS.event calls being present.'));
```

Comment:
Agreed; I sort of regret not refactoring this earlier. Fortunately, this should be the last PR before getting out a general browser test cleanup PR.

I was originally thinking of introducing something like `await testHarness.ensureNoFSEventCalls()` that would reject a promise like we have here, but your comment makes me wonder if it would be better to update `testHarness.popEvent()` to return `undefined` if the timeout is reached instead of rejecting the promise. This would let us:

```
expect(await testHarness.popEvent(500)).to.be.undefined;
```

WDYT?

Reply:
I like the idea of returning undefined here instead, it makes the expect look really neat 👍 


this doesn't require action and is more for my understanding: does a rejected promise throw an exception and if so could we catch it? sample code:

```
try {
  await testHarness.popEvent(500)
  expect.fail()
} catch (exception) {
   expect(exception).is.notNull
}
```